dianna/methods/shap.py

```diff
@@ -1,21 +1,223 @@
+        max_num_iter = kwargs.get("max_num_iter", 10)
+        spacing = kwargs.get("spacing", None)
+        convert2lab = kwargs.get("convert2lab", None)
+        enforce_connectivity = kwargs.get("enforce_connectivity", True)
+        min_size_factor = kwargs.get("min_size_factor", 0.5)
+        max_size_factor = kwargs.get("max_size_factor", 3)
+        slic_zero = kwargs.get("slic_zero", False)
+        start_label = kwargs.get("start_label", 1)
+        mask = kwargs.get("mask", None)
+
+        # first check the dimension of input_data
+        if input_data.ndim != 3:
+            raise IOError(
+                "The input image must follow the required shape [height, width, color_channels] or [color_channels, height, width]"
+            )
+
+        # call the segment method to create segmentation of input image
+        self.image_segments = self._segment_image(
+            input_data,
+            n_segments,
+            compactness,
+            max_num_iter,
+            sigma,
+            spacing,
+            convert2lab,
+            enforce_connectivity,
+            min_size_factor,
+            max_size_factor,
+            slic_zero,
+            start_label,
+            mask,
+            channel_axis_first,
+        )
+
+        # call the Kernel SHAP explainer
+        explainer = shap.KernelExplainer(self._runner, np.zeros((1, n_segments)))
+
+        with warnings.catch_warnings():
+            # avoid warnings due to version conflicts
+            warnings.simplefilter("ignore")
+            shap_values = explainer.shap_values(
+                np.ones((1, n_segments)), nsamples=nsamples
+            )
+
+        return shap_values, self.image_segments
+
+    def _mask_image(
+        self, features, segmentation, image, background=None, channel_axis_first=False
+    ):
+        """Define a function that depends on a binary mask representing
+           if an image region is hidden.
+
+        Args:
+            features (np.ndarray): A matrix of samples (# samples x # features)
+                                   on which to explain the model's output.
+            segmentation (np.ndarray): Image segmentations generated by
+                                       the function _segment_image
+            image (np.ndarray): Image to be explained
+            background (int): Background color for the masked image
+            channel_axis (int): See the function explain_image
         """
-        self.model = model
+        # if the image shape is [color_channels, height, width]
+        if channel_axis_first:
+            image = np.transpose(image, (1, 2, 0))
+        # check the background color
+        if background is None:
+            background = image.mean((0, 1))
+
+        # Create an empty 4D array
+        out = np.zeros(
+            (features.shape[0], image.shape[0], image.shape[1], image.shape[2])
+        )
+
+        for i in range(features.shape[0]):
+            out[i, :, :, :] = image
```

Comment:
Small suggestion that I think looks a little bit nicer: `out[i, ...] = image` would also work. Actually just `out[i] = image` works too.

Reply:
Thinking about it, perhaps the out array can be generated with something like `np.tile`. Then we may not need the double for loop anymore, but could do everything with np.where etc.

Reply:
Thanks for the comment. Just change it to `out[i] = image`, but didn't find a way to simplify the code due to the if statement nested in the loop and the mismatch of the shape. Any thoughts about that? Otherwise I would just leave it like this, though it is not perfect, for sure.